module/zfs/ddt.c

```diff
@@ -822,6 +864,177 @@ ddt_key_compare(const void *x1, const void *x2)
+ * if there is nothing stored, select based on pool config.
+ */
+static int
+ddt_configure(ddt_t *ddt)
+{
+	spa_t *spa = ddt->ddt_spa;
+	char name[DDT_NAMELEN];
+	int error;
+
+	ASSERT3U(spa_load_state(spa), !=, SPA_LOAD_CREATE);
+
+	boolean_t fdt_enabled =
+	    spa_feature_is_enabled(spa, SPA_FEATURE_FAST_DEDUP);
+	boolean_t fdt_active =
+	    spa_feature_is_active(spa, SPA_FEATURE_FAST_DEDUP);
+
+	/*
+	 * First, look for the global DDT stats object. If its not there, then
+	 * there's never been a DDT written before ever, and we know we're
+	 * starting from scratch.
+	 */
+	error = zap_lookup(spa->spa_meta_objset, DMU_POOL_DIRECTORY_OBJECT,
+	    DMU_POOL_DDT_STATS, sizeof (uint64_t), 1,
+	    &spa->spa_ddt_stat_object);
+	if (error != 0) {
+		if (error != ENOENT)
+			return (error);
+		goto new;
+	}
+
+	if (fdt_active) {
+		/*
+		 * Now look for a DDT directory. If it exists, then it has
+		 * everything we need.
+		 */
+		snprintf(name, DDT_NAMELEN, DMU_POOL_DDT_DIR,
+		    zio_checksum_table[ddt->ddt_checksum].ci_name);
+
+		error = zap_lookup(spa->spa_meta_objset,
+		    DMU_POOL_DIRECTORY_OBJECT, name, sizeof (uint64_t), 1,
+		    &ddt->ddt_dir_object);
+		if (error == 0) {
+			ASSERT3U(spa->spa_meta_objset, ==, ddt->ddt_os);
+
+			error = zap_lookup(ddt->ddt_os, ddt->ddt_dir_object,
+			    DDT_DIR_VERSION, sizeof (uint64_t), 1,
+			    &ddt->ddt_version);
+			if (error != 0)
+				return (error);
+
+			error = zap_lookup(ddt->ddt_os, ddt->ddt_dir_object,
+			    DDT_DIR_FLAGS, sizeof (uint64_t), 1,
+			    &ddt->ddt_flags);
+			if (error != 0)
+				return (error);
+
+			if (ddt->ddt_version != DDT_VERSION_FDT) {
+				zfs_dbgmsg("ddt_configure: spa=%s ddt_dir=%s "
+				    "unknown version %llu", spa->spa_name,
+				    name, (u_longlong_t)ddt->ddt_version);
+				return (SET_ERROR(EINVAL));
+			}
+
+			if ((ddt->ddt_flags & ~DDT_FLAG_MASK) != 0) {
+				zfs_dbgmsg("ddt_configure: spa=%s ddt_dir=%s "
+				    "version=%llu unknown flags %llx",
+				    spa->spa_name, name,
+				    (u_longlong_t)ddt->ddt_flags,
+				    (u_longlong_t)ddt->ddt_version);
+				return (SET_ERROR(EINVAL));
+			}
+
+			return (0);
+		}
+		if (error != ENOENT)
+			return (error);
+	}
+
+	/* Any object in the root indicates a traditional setup. */
+	for (ddt_type_t type = 0; type < DDT_TYPES; type++) {
+		for (ddt_class_t class = 0; class < DDT_CLASSES; class++) {
+			ddt_object_name(ddt, type, class, name);
+			uint64_t obj;
+			error = zap_lookup(spa->spa_meta_objset,
+			    DMU_POOL_DIRECTORY_OBJECT, name, sizeof (uint64_t),
+			    1, &obj);
+			if (error == ENOENT)
+				continue;
+			if (error != 0)
+				return (error);
+
+			ddt->ddt_version = DDT_VERSION_LEGACY;
+			ddt->ddt_flags = ddt_version_flags[ddt->ddt_version];
+			ddt->ddt_dir_object = DMU_POOL_DIRECTORY_OBJECT;
+
+			return (0);
+		}
+	}
+
+new:
+	/* Nothing on disk, so set up for the best version we can */
+	if (fdt_enabled) {
+		ddt->ddt_version = DDT_VERSION_FDT;
```

Comment:
Here if nothing is on disk you set version to DDT_VERSION_FDT, while in ddt_sync_table() when last record is deleted you clear the version to DDT_VERSION_UNCONFIGURED.  Is it not very consistent, or I am missing something?

Reply:
I have no questions for the case of the last entry removal.  My question is about pool import case, when if nothing is found on disk you may set version to FDT, respecting the feature enabled in that moment.  But if the feature is later disabled, all the tables with version set to FDT will remain as such, while without anything on disk they could be made LEGACY on first touch.  I think ddt_configure() could get a flag, specifying whether it is called on import, that means it may leave the version UNCONFIGURED, or it is a first touch, and the choice must be made at this point.

Reply:
Got it. Turns out we can just do `ddt_configure()` at first write, not first lookup. That is, `ddt_sync_table()`, not `ddt_lookup()`. [[change]](https://github.com/openzfs/zfs/compare/08aa1eb178966e013b8bdda099ab115fff5fa778..e7636dd6722e478afb6d199fc9f3ebd001cc5a14).

I wrote some basic tests for the two types for create and import, and for upgrade. I'll massage them into ZTS shapes now and add them to this PR.

Reply:
Ahh, turns out we can't, and rebasing the stack above this PR has reminded me why. The choice of FDT & feature flags decide the in-memory format of the entry. So it does have to be done in `ddt_lookup()`, basically the first time we allocate an entry. So I'll rework a little; mostly moving the configure call back to lookup, and activating the feature right there, so the choice is stable.

Reply:
Can't do that either; I don't have a valid `dmu_tx` in `ddt_lookup()` to hang the feature bump from. Bah.

So yeah, I'll be doing something like your suggestion then. It felt a little awkward, but it'll probably be fine, and maybe I should just listen next time :D

Reply:
Alright, last push gets it. What a day :/